internal/pkg/agent/protection/action.go

```
@@ -0,0 +1,96 @@
+	ErrNonMatchingActionID    = errors.New("non-matching action id")
+	ErrInvalidSignedDataValue = errors.New("invalid signed data value")
+	ErrInvalidSignatureValue  = errors.New("invalid signature value")
+)
+
+type fleetActionWithAgents struct {
+	ActionID         string          `json:"action_id"` // Note the action_id here, since the signed action uses action_id for id
+	ActionType       string          `json:"type,omitempty"`
+	InputType        string          `json:"input_type,omitempty"`
+	Timestamp        string          `json:"@timestamp"`
+	ActionExpiration string          `json:"expiration,omitempty"`
+	ActionStartTime  string          `json:"start_time,omitempty"`
+	Timeout          int64           `json:"timeout,omitempty"`
+	Data             json.RawMessage `json:"data,omitempty"`
+	Agents           []string        `json:"agents"`
+}
+
+// ValidateAction validates action signature, checks the signed payload action id matches the action id, checks the agent id match
+func ValidateAction(a fleetapi.ActionApp, signatureValidationKey []byte, agentID string) (fleetapi.ActionApp, error) {
+	// Nothing to validate if not signed
+	if a.Signed == nil {
+		return a, nil
+	}
+
+	data, err := base64.StdEncoding.DecodeString(a.Signed.Data)
+	if err != nil {
+		//nolint:errorlint // WAD: unfortunately two errors wrapping is only available in Go 1.20
+		return a, fmt.Errorf("%w: %v", ErrInvalidSignedDataValue, err)
+	}
+
+	signature, err := base64.StdEncoding.DecodeString(a.Signed.Signature)
+	if err != nil {
+		//nolint:errorlint // WAD: unfortunately two errors wrapping is only available in Go 1.20
+		return a, fmt.Errorf("%w: %v", ErrInvalidSignatureValue, err)
+	}
+
+	if len(signatureValidationKey) != 0 {
+		// Validate signature
+		err = ValidateSignature(data, signature, signatureValidationKey)
+		if err != nil {
+			return a, err
+		}
+	}
+
+	// Deserialize signed action data if it's a valid JSON
+	if json.Valid(data) {
```

Should you return an error when this evaluates to false?